pkg/controller/atlasdeployment/atlasdeployment_controller.go

```diff
@@ -95,61 +94,112 @@ func (r *AtlasDeploymentReconciler) Reconcile(context context.Context, req ctrl.
+				result = workflow.Terminate(workflow.Internal, err.Error())
+				log.Errorw("Failed to remove finalizer", "error", err)
+				return result.ReconcileResult(), nil
+			}
+		}
 		return workflow.OK().ReconcileResult(), nil
 	}
 
-	ctx := customresource.MarkReconciliationStarted(r.Client, deployment, log)
+	customContext := customresource.MarkReconciliationStarted(r.Client, deployment, log)
 	log.Infow("-> Starting AtlasDeployment reconciliation", "spec", deployment.Spec, "status", deployment.Status)
-	defer statushandler.Update(ctx, r.Client, r.EventRecorder, deployment)
+	defer statushandler.Update(customContext, r.Client, r.EventRecorder, deployment)
 
 	if err := validate.DeploymentSpec(deployment.Spec); err != nil {
 		result := workflow.Terminate(workflow.Internal, err.Error())
-		ctx.SetConditionFromResult(status.ValidationSucceeded, result)
+		customContext.SetConditionFromResult(status.ValidationSucceeded, result)
 		return result.ReconcileResult(), nil
 	}
-	ctx.SetConditionTrue(status.ValidationSucceeded)
+	customContext.SetConditionTrue(status.ValidationSucceeded)
 
 	project := &mdbv1.AtlasProject{}
-	if result := r.readProjectResource(deployment, project); !result.IsOk() {
-		ctx.SetConditionFromResult(status.DeploymentReadyType, result)
+	if result := r.readProjectResource(ctx, deployment, project); !result.IsOk() {
+		customContext.SetConditionFromResult(status.DeploymentReadyType, result)
 		return result.ReconcileResult(), nil
 	}
 
 	connection, err := atlas.ReadConnection(log, r.Client, r.GlobalAPISecret, project.ConnectionSecretObjectKey())
 	if err != nil {
 		result := workflow.Terminate(workflow.AtlasCredentialsNotProvided, err.Error())
-		ctx.SetConditionFromResult(status.DeploymentReadyType, result)
+		customContext.SetConditionFromResult(status.DeploymentReadyType, result)
 		return result.ReconcileResult(), nil
 	}
-	ctx.Connection = connection
+	customContext.Connection = connection
 
 	atlasClient, err := atlas.Client(r.AtlasDomain, connection, log)
 	if err != nil {
 		result := workflow.Terminate(workflow.Internal, err.Error())
-		ctx.SetConditionFromResult(status.DeploymentReadyType, result)
+		customContext.SetConditionFromResult(status.DeploymentReadyType, result)
 		return result.ReconcileResult(), nil
 	}
-	ctx.Client = atlasClient
+	customContext.Client = atlasClient
 
 	// Allow users to specify M0/M2/M5 deployments without providing TENANT for Normal and Serverless deployments
 	r.verifyNonTenantCase(deployment)
 
+	if deployment.GetDeletionTimestamp().IsZero() {
+		if !haveFinalizer(deployment, finalizer) {
+			deployment.SetFinalizers(append(deployment.GetFinalizers(), finalizer))
+			if err := r.Client.Update(ctx, deployment); err != nil {
+				result = workflow.Terminate(workflow.Internal, err.Error())
+				customContext.SetConditionFromResult(status.DeploymentReadyType, result)
+				return result.ReconcileResult(), nil
+			}
+		}
+	}
+
+	if !deployment.GetDeletionTimestamp().IsZero() {
```

Comment:
Can be replaced with `else` by joining with previous `if` statement?

Reply:
I think this is the same style as for projects. Sometimes it's considered a good practice to avoid `else` statement as it requires the developer to remember what was in the first `if` statement.

In that case I think both variants are Ok as the blocks aren't big.